### ISSUE TITLE
Problem with the way Caddy serves multiple certificates

### ISSUE THREAD

**Issue by whalehub** [mholt/caddy#2334]
### 1. What version of Caddy are you using (`caddy -version`)?

Caddy 0.11.0 (+22dfb14 Sun Nov 04 19:32:32 UTC 2018) (unofficial)

### 2. What are you trying to do?

I'm trying to fetch a certificate from my domain app01.domain.tld by using the following shell script.

```
#!/bin/bash
#
# usage: sh script.sh app01.domain.tld 443
#
REMHOST=$1
REMPORT=${2:-443}

echo |\
openssl s_client -connect ${REMHOST}:${REMPORT} 2>&1 |\
sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p'

```

### 3. What is your entire Caddyfile?
```
(ecc-tls) {

   tls {
       dns gandiv5
       ciphers ECDHE-ECDSA-AES256-GCM-SHA384 ECDHE-ECDSA-AES128-GCM-SHA256 ECDHE-ECDSA-WITH-CHACHA20-POLY1305
       key_type p256
       must_staple
    }
}

app01.domain.tld {

   import ecc-tls
}

app02.domain.tld {

   import ecc-tls
}

app03.domain.tld {

   import ecc-tls
}

```

### 4. How did you run Caddy (give the full command and describe the execution environment)?

```
caddy -agree -http2 -quic -log /opt/caddy/log -root /opt/caddy/webroot -conf /opt/caddy/Caddyfile
```

I'm running Caddy on Debian 9.5 (stretch).

### 5. Please paste any relevant HTTP request(s) here.

N/A

### 6. What did you expect to see?

The certificate of app01.domain.tld.

### 7. What did you see instead (give full error messages and/or log)?

The certificate of either app01, app02 or app03.domain.tld. The certificate changes each time you run the shell script.

### 8. How can someone who is starting from scratch reproduce the bug as minimally as possible?

1. Install Caddy.
2. Set up three or more subdomains on your own domain.
3. Copy my Caddyfile and adjust it as needed to fetch a certificate for each subdomain.
4. Use the provided shell script to fetch the certificate of one of the subdomains.

This issue does not affect sites which are not served with Caddy. To verify this, run the same shell script on such a domain and observe that the certificate in the output is the same every time you run it and that it is the correct one for the domain.

The practical relevance of this problem is that the remote control software MeshCentral cannot fetch the correct certificate from a subdomain that is served with Caddy because it [receives a different certificate](https://github.com/Ylianst/MeshCentral/issues/30#issuecomment-435291069) each time it tries to fetch it.

**Comment by ghoeffner** [comment]
Related to #1303 👍

**Comment by whalehub** [comment]
@tobya No, specifying :443 at the end of the domains sadly didn't fix my issue. The certificate still changes each time I run the shell script.

As for question 5, I'm not sure where I can get the HTTP request from. I'm using openssl to request the certificate from the domain (see shell script), so I'm not seeing the actual HTTP request in the terminal.

**Comment by tobya** [comment]
Can you specify how you are calling your shell script.  Can you use an actual call.

**Comment by whalehub** [comment]
Just for future reference: The reason why the fetched certificate changed each time is because I wasn't telling OpenSSL to use SNI. The following shell script can be used to fetch the correct certificate each time:

```
#!/bin/bash
#
# usage: sh script.sh app01.domain.tld 443
#
REMHOST=$1
REMPORT=${2:-443}

echo |\
openssl s_client -servername ${REMHOST} -connect ${REMHOST}:${REMPORT} 2>&1 |\
sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p'

```

**Comment by tobya** [comment]
@whalehub thanks for the update.  The underlying issue when you didnt specify SNI is fixed with  #2339.

**Comment by whalehub** [comment]
@tobya I can confirm that OpenSSL now returns a TLS alert instead of a random certificate when attempting to fetch the certificate without using SNI.
